examples/benchmarks/other/nbe.effekt

```diff
@@ -0,0 +1,151 @@
+
+/// lambda term
+type Term {
+  Abs(n: Name, t: Term)
+  App(a: Term, b: Term)
+  Var(n: Name)
+}
+
+/// lambda term in normal domain
+type Value {
+  VNeu(neu: Neutral)
+  VClo(n: Name, t: Term, env: Map[Name, Value])
+}
+
+/// lambda term in neutral domain (not yet reducible)
+type Neutral {
+  NVar(n: Name)
+  NApp(a: Neutral, b: Value)
+}
+
+/// evaluate a single term without going into abstractions
+def eval(env: Map[Name, Value], t: Term): Value = t match {
+  case Abs(n, t) => VClo(n, t, env)
+  case App(a, b) => apply(eval(env, a), eval(env, b))
+  case Var(n)    => env.getOrElse(n) { VNeu(NVar(n)) }
+}
+
+/// apply terms in their normal domain
+/// this does the actual substitution via lookup handling (or bubbling up respectively)
+def apply(a: Value, b: Value): Value = a match {
+  case VNeu(neu)       => VNeu(NApp(neu, b))
+  case VClo(n, t, env) => eval(env.put(n, b), t)
+}
+
+/// reflect variable name to the neutral domain
+def reflect(n: Name): Value = VNeu(NVar(n))
+
+/// convert terms to their normal form (in term domain)
+def reify(v: Value): Term / fresh = v match {
+  case VNeu(NVar(n)) => Var(n)
+  case VNeu(NApp(a, b)) => App(reify(VNeu(a)), reify(b))
+  case _ => {
+    val n = do fresh()
+    Abs(n, reify(apply(v, reflect(n))))
+  }
+}
+
+/// strong normalization of the term
+def normalize(t: Term): Term = {
+  var i = 0
+  try reify(eval(empty[Name, Value](compareInt), t))
+  with fresh { i = i + 1; resume(i) }
+}
+
+/// parse named term from BLC stream (de Bruijn)
+def parse(): Term / { read[Bool], stop } = {
+  def go(): Term / { fresh, lookup } =
+    (do read[Bool](), do read[Bool]) match {
+      case (false, false) => {
+        val n = do fresh()
+        Abs(n, try go() with lookup { i =>
+          resume(if (i == 0) n else do lookup(i - 1))
+        })
+      }
+      case (false, true ) => App(go(), go())
+      case (true , false) => Var(do lookup(0))
+      case (true , true ) => {
+        var i = 1
+        while (do read[Bool]()) i = i + 1
+        Var(do lookup(i))
+      }
+    }
+
+  var i = 0
+  try go()
+  with fresh { i = i + 1; resume(i) }
+  with lookup { n =>
+    println("error: free variable " ++ n.show)
+    exit(1)
+  }
+}
+
+/// helper function for pretty string interpolation of terms
+def pretty { s: () => Unit / { literal, splice[Term] } }: String = {
+  with stringBuffer
+
+  try { s(); do flush() }
+  with literal { l => resume(do write(l)) }
+  with splice[Term] { t =>
+    t match {
+      case Abs(n, t) => do write("λ" ++ n.show ++ pretty".${t}")
+      case App(a, b) => do write(pretty"(${a} ${b})")
+      case Var(v)    => do write(v.show)
+    }
+    resume(())
+  }
+}
+
+/// convert char stream to bit stream, skipping non-bit chars
+def bits { p: () => Unit / { read[Bool], stop } }: Unit / read[Char] =
+  try exhaustively { p() }
+  with read[Bool] {
+    with exhaustively
+    val c = do read[Char]()
+    if (c == '0') resume { false }
+    if (c == '1') resume { true  }
+  }
```

Comment:
Cool!

Reply:
We should really have something like this in the standard library though! Something like a `map` or `filter` @phischu